Added reducer and tests for updateMatchIdToMatchState

herbertli opened the pull request:
- added action updateMatchIdToMatchState to reducer
- modified action setMatchList to add initial match state to matchIdToMatchState for new matches
- added tests in index.test.ts for updateMatchIdToMatchState
- modified tests for setMatchList to create initial match states when necessary
- added invariant checks to make sure every matchId in matchesList exists in matchIdToMatchState
- added default for matchIdToMatchState in store

I'm not sure if this is what you meant by invariance checks, I'm also not sure if I overcomplicated things  a bit (e.g. it's not necessary to check that all matchIds exist in matchIdToMatchState)

yoav-zibin commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,50 @@ function checkStoreInvariants(state: StoreState) {
   );
+
+  checkCondition(
+    'every matchId in matchesList is also present in matchIdToMatchState',
```
We actually need the other direction :)
Match state will be loaded lazily.


  // Match state will be loaded lazily, i.e.,
  // not every match in matchesList will have an entry in matchIdToMatchState.
  matchIdToMatchState: MatchIdToMatchState;

yoav-zibin commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,50 @@ function checkStoreInvariants(state: StoreState) {
+      if (e.matchId in matchIdToMatchState) {
+        newMatchIdToMatchState[e.matchId] = matchIdToMatchState[e.matchId];
+      } else {
+        newMatchIdToMatchState[e.matchId] = {};
```
nicely done :)

herbertli commented:
Updated: Reverse invariance check, all matchIds in matchIdToMatchState should correspond to a match in matchesList

yoav-zibin commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
   } else if (action.setMatchesList) {
-    return { ...state, matchesList: action.setMatchesList };
+    let { matchesList, matchIdToMatchState, ...rest } = state;
+    let newMatchIdToMatchState = {};
```
how does this type check? don't you need explicit types here?

yoav-zibin commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
+      if (e.matchId in matchIdToMatchState) {
+        newMatchIdToMatchState[e.matchId] = matchIdToMatchState[e.matchId];
+      } else {
+        newMatchIdToMatchState[e.matchId] = {};
```
delete this line.
matchIdToMatchState[someID] should tell you if we have state or not.
Setting {} is confusing :)

yoav-zibin commented on src/stores/defaults.ts:
```diff
@@ -51,7 +51,10 @@ export const storeStateDefault: StoreState = {
   ],
   currentMatchIndex: 1,
-  matchIdToMatchState: {},
+  matchIdToMatchState: {
```
revert

apsdehal replied:
Yes, you don't want these defaults to change. You want to use tests to test your reducer.

yoav-zibin commented on src/reducers/index.test.ts:
```diff
@@ -71,7 +73,20 @@ it('setMatchesList', () => {
   };
   let initialState = storeStateDefault;
-  const expectedState = { ...storeStateDefault, matchesList: matchesList };
+  let { matchIdToMatchState } = initialState;
```
revert this test. This test shouldn't need any changes.

Please add another test where the state has a match in matchesList and matchIdToMatchState, and then we call setMatchesList with an empty list, so that should delete from both.

herbertli replied:
for that test, should I reset the currentMatchIndex to like -1 or something? Or else the bounds check will fail

yoav-zibin replied:
Perfect, you found a bug :)
currentMatchIndex may need to be updated when dispatching setMatchesList
Yes, make sure we have a test that covers that.
Thanks!!

yoav-zibin commented on src/reducers/index.test.ts:
```diff
@@ -107,4 +122,32 @@ it('setCurrentMatchIndex', () => {
+  let { matchIdToMatchState, ...rest } = initialState;
+  var someId = '';
+  var count = 0;
+  // select a random matchId to update
```
please make tests deterministic.

apsdehal replied:
Test for all updates maybe in a loop.

yoav-zibin commented:
Note that "This branch has conflicts that must be resolved" (I just merged another PR)

apsdehal commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
+    let { matchesList, matchIdToMatchState, ...rest } = state;
+    let newMatchIdToMatchState = {};
+    action.setMatchesList.forEach(e => {
+      if (e.matchId in matchIdToMatchState) {
```
Safer to add `matchIdToMatchState.hasOwnProperty(e.matchId)`. Though, this should also work here.
Some ES6 syntactic sugar, but decreases readability (so I don't recommend it):

```
newMatchIdToMatchState[e.matchId] = {
	...((e.matchId in matchIdToMatchState) && matchIdToMatchState[e.matchId])
}

apsdehal commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
+  checkCondition(
+    'every matchId in matchIdToMatchState is also present in matchesList',
+    Object.keys(state.matchIdToMatchState).reduce(
+      (a, e) =>
```
Better to give them readable names then a and e (even though they prev and curr)

apsdehal commented on src/reducers/index.ts:
```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
-
+    return { ...state, currentMatchIndex: action.setCurrentMatchIndex };
+  } else if (action.updateMatchIdToMatchState) {
+    let { matchIdToMatchState, ...rest } = state;
```
Remove this line and directly pass `...state` instead of rest and use `state[matchIdToMatchState` in second line.

I'm being picky here, both are ok though.

yoav-zibin commented on src/reducers/index.ts:
```diff
@@ -67,18 +69,28 @@ function reduce(state: StoreState, action: Action) {
-        newMatchIdToMatchState[e.matchId] = {};
       }
     });
+    if (
```
You uncovered another bug of mine (thanks again!)

Change above:
checkCondition(
     'currentMatchIndex is in range',
     state.currentMatchIndex >= -1 &&
       state.currentMatchIndex <= state.matchesList.length
   );

to:

checkCondition(
     'currentMatchIndex is in range',
     state.currentMatchIndex >= -1 &&
       state.currentMatchIndex < state.matchesList.length
   );

I don't know what I was thinking :)
Allowing currentMatchIndex === matchesList.length will result in an out-of-bound access...

Extract it into a function:
function isInRange(currentMatchIndex, matchesList) {
  return currentMatchIndex >= -1 &&
       currentMatchIndex < matchesList.length;
}

And reuse that function here and in the invariant check.
If we're not in range, then set currentMatchIndex to -1  (not to ction.setMatchesList.length)

herbertli commented:
- isInRange function
- updated invariant checks
- updated setMatchesList to set currentMatchIndex to -1 if the index becomes out of bounds
- updated tests for setMatchesList to correct output

yoav-zibin commented:
Thanks Herbert!
